lib/src/builder.dart

```diff
@@ -167,25 +167,32 @@ $rawOutput
     // `Future`, which is needed when overriding some methods which return
     // `FutureOr`.
     final typeVisitor = _TypeVisitor(entryLib.typeProvider.futureDynamicType);
-    final seenTypes = <analyzer.InterfaceType>{};
+    final seenTypes = <analyzer.DartType>{};
     final librariesWithTypes = <LibraryElement>{};
 
-    void addTypesFrom(analyzer.InterfaceType type) {
+    void addTypesFrom(analyzer.DartType type) {
       // Prevent infinite recursion.
       if (seenTypes.contains(type)) {
+        if (type.alias != null) {
+          // To check for duplicate typdefs that have different names
+          type.alias!.element.accept(typeVisitor);
+        }
         return;
       }
       seenTypes.add(type);
-      librariesWithTypes.add(type.element.library);
-      type.element.accept(typeVisitor);
-      if (type.alias != null) type.alias!.element.accept(typeVisitor);
-      // For a type like `Foo<Bar>`, add the `Bar`.
-      type.typeArguments
-          .whereType<analyzer.InterfaceType>()
-          .forEach(addTypesFrom);
-      // For a type like `Foo extends Bar<Baz>`, add the `Baz`.
-      for (final supertype in type.allSupertypes) {
-        addTypesFrom(supertype);
+      librariesWithTypes.addAll([
+        if (type.element?.library != null) type.element!.library!,
```

Comment:
I think this can read better if we hoist these properties out with `if case`, instead of `addAll` and `if`:

```dart
if (type.element?.library case var library?) {
  librariesWithTypes.add(library);
}
if (type.alias?element.library case var library?) {
  librariesWithTypes.add(library);
}
```

The `var library?` syntax is still kind of weird to me, but I'm trying to use it more. Basically "if that is non-null then extract it to a variable named `library`."

Reply:
Agree. Big fan of the `case` in `if` checks too.
It's a shame it isn't a well known feature.